### DISCUSSION TITLE
Jotai with Remix

### DISCUSSION THREAD

**Discussion by spacepolice10** [pmndrs/jotai#2513] [Q&A]
Hi there, 

Has anyone ever tried to use Jotai with Remix.run? I tried to use it with SSR but it always fails to hydrate correctly. No matter if I use Provider with useHydrateAtoms or not all the Jotai related content (data that I try to read from atoms) appears with delay and break hydration process. 

For example. I have a server loader that returns userID

```tsx
export const loader = async ( ) => userId
```

Than I want to use it data before hydration

```tsx
export function Component () {
  const data = useLoaderData()
  return <Provider>
    <Hydration userId={data?.userId}>
       {children}
    </Hydration>
  </Provider>
}
```

```tsx
export function Hydration({children) {
 useHydrateAtoms([[userIdAtom, userId]])
 return children
}
```

Whenever I try to use this code I receive hydration error because data from the server kinda delayed. Even if I reload the page and see that `data.userId` shows correct updated data, the atom `userIdAtom` is stale. 

Let's say I logged in as **userId == 0,** if I open a new tab and log in as **userId == 1**, Jotai atom will show old data, it will say that I am **userId == 0**. EVEN if I reload the page. Even though the loader itself will return **userId == 1** atoms return previoys data. Anyone with experience in such cases? 

P.S. Using forceHydrate eliminates this problem but generates new ones, strongly discouraged to use & should not be needed in this situation anyway

**Comment by dai-shi** [comment]
If the data is delayed, you would need to render twice:
```js
export function Hydration({ userId, children }) {
  useHydrateAtoms([[userIdAtom, userId]])
  const setUerId = useSetAtom(userIdAtom)
  useEffect(() => {
    setUserId[userId]
  }, [setUserId, userId])
 return children
}
```

**Reply by spacepolice10** [reply]
It is actually not. I tried to log the data. Afaik, the page is never rendered before `loader` is ready. And when `loader` is ready it provides fresh data. When I log the data in Hydration component, it is also fresh. But when I read the same piece of data down the React tree from atom, it is stale. I also hydrate React queryClient and it works fine. That is why I decided to start discussion here and not in Remix repo.
Also I am concerned about the staleness of data. I can't understand why there's ANY data even if I reload the page (Jotai seems to be client-only entitiy). Maybe, I have somehow bundled and cached it with within server chunk. I'll learn more. Also, I will try to provide.
Anyway, thank you for advice. I will provide minimal repro asap, maybe it will make things more clear. Thanks in advance again

**Reply by spacepolice10** [reply]
Here it is 

https://github.com/spacepolice10/remix-jotai-hydr-bugs

When I try to switch user and then reload the page, I receive hydration mismatch error because atom stores incorrect (stale, selected previously) data. Even though loader returns fresh info about the user

P.S. Probably it is too early to call it a bug, it is just something that came up into my mind )

**Reply by spacepolice10** [reply]
~Ok. I found out that hydration mismatch happens when I use `store` API or jotai-tanstack-query. If I pass `createStore` to **Provider** and then try to access any data stored in Jotai, I receive an error. If I use **Provider** only & instead of exploiting `store` methods read state with hooks, everything seem to be fine. If I understand correctly, Jotai doesn't work correctly server-side (especially functions like `createStore` & `getDefaultStore`) and heavily depends on something client-side related. Is it expected to be so? Should I reconsider the way I use Jotai and stop using `store` in such cases?~

It feels like I found a way to avoid my problem. What I did:

1. created separate store using `clientOnly$` utility, so it is never referenced in server bundle
2. passed made-up store to **Provider** AND to `useHydrateAtoms` function

That's it. Everything plays nicely together & no hydration mismatch present
